### PR TITLE
Allow dotnet watch to refresh on changes to .razor files in reference…

### PR DESCRIPTION
**Pull request by pranavkm**
…d projects

dotnet-watch expects full paths to files to watch in all added items. For files in the current project, this
happens to work. However this does not work very well for files in referenced projects. Using the FullPath metadata
does not work during global evaluation, but we're able to do this in a target.

This change adds extensibility to the dotnet-watch tool that allows calling a target as part of "watch" evaluation.

Fixes https://github.com/dotnet/aspnetcore/issues/22219

### CODE REVIEW

File: src/Tools/dotnet-watch/test/MsBuildFileSetFactoryTest.cs [hunk 1]
```diff
@@ -276,6 +277,53 @@ public async Task ProjectReferences_Graph()
...
+        {
+            // Arrange
+            var razorSdkTargetsPath = GetType().Assembly.GetCustomAttributes<AssemblyMetadataAttribute>()
+                .First(f => f.Key == "RazorSdkPath").Value;
+
+            _tempDir
+                .SubDir("src")
+                    .SubDir("MyLibrary")
+                        .WithCSharpProject("MyLibrary", out var proj3, sdk: "Microsoft.NET.Sdk.Razor")
+                        .WithProperty("CustomCollectWatchItems", "$(CustomCollectWatchItems);_RazorSdkCustomCollectWatchItems")
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/test/MsBuildFileSetFactoryTest.cs:291]: We cannot reference the Sdk.props in this repo (same deal as https://github.com/dotnet/aspnetcore/issues/24278). Changing this in the project instead to simulate this.

File: src/Tools/dotnet-watch/test/TestProjects/KitchenSink/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+.net
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/test/TestProjects/KitchenSink/.gitignore:1]: Building the project constantly adds build artifacts that are not gitignored

File: src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets [hunk 3]
```diff
@@ -870,6 +864,14 @@ Copyright (c) .NET Foundation. All rights reserved.
     </PropertyGroup>
   </Target>
 
+  <Target Name="_RazorSdkCustomCollectWatchItems">
+    <ItemGroup>
+      <Watch Include="%(Content.FullPath)" Condition="'%(Content.Extension)' == '.razor' AND '%(Content.Watch)' != 'false'" />
```
Comment (captainsafia) [inline src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets:869]: Is there a reason that this isn't `Content.Watch == 'true'`?

As in, I would expect `Content.Watch` to be an empty value if it is not set.

Comment (pranavkm) [reply src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets:869]: > As in, I would expect Content.Watch to be an empty value if it is not set.

Yup. By default, this value is never set. You can opt individual files out of watch by setting them to `false`

**Review by captainsafia — APPROVED** [review]

**Comment by SteveSandersonMS** [comment]
Can we include .razor.css too?

**Comment by Pilchie** [comment]
Approved for RC1.

**Comment by pranavkm** [comment]
> Can we include .razor.css too?

👍

File: src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets (hunk 3 above)
Comment (pranavkm) [reply src/Razor/Microsoft.NET.Sdk.Razor/src/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets:869]:
```suggestion
      <Watch Include="%(Content.FullPath)" Condition="'%(Content.Extension)' == '.razor' AND '%(Content.Watch)' != 'false'" />
      <Watch Include="%(Content.FullPath)" Condition="'%(Content.Extension)' == '.razor.css' AND '%(Content.Watch)' != 'false'" />
```

**Comment by mkArtakMSFT** [comment]
Thanks @pranavkm !

**Pull request merged**